iceoryx_posh/test/moduletests/test_popo_subscriber_port.cpp

```diff
@@ -0,0 +1,275 @@
+    m_sutUserSideSingleProducer.unsubscribe();
+
+    auto maybeCaproMessage = m_sutRouDiSideSingleProducer.getCaProMessage();
+
+    EXPECT_TRUE(maybeCaproMessage.has_value());
+    auto caproMessageResponse = maybeCaproMessage.value();
+    EXPECT_THAT(caproMessageResponse.m_type, Eq(iox::capro::CaproMessageType::UNSUB));
+    EXPECT_THAT(caproMessageResponse.m_serviceDescription, Eq(iox::capro::ServiceDescription("x", "y", "z")));
+}
+
+TEST_F(SubscriberPort_test, unsubscribeInSubscribedResultsInUnsubscribeRequested)
+{
+    m_sutUserSideSingleProducer.subscribe();
+    m_sutRouDiSideSingleProducer.getCaProMessage(); // only RouDi changes state
+    iox::capro::CaproMessage caproMessage(iox::capro::CaproMessageType::ACK,
+                                          iox::capro::ServiceDescription("x", "y", "z"));
+    m_sutRouDiSideSingleProducer.dispatchCaProMessage(caproMessage);
+    m_sutUserSideSingleProducer.unsubscribe();
+    m_sutRouDiSideSingleProducer.getCaProMessage(); // only RouDi changes state
+
+    const auto subscriptionState = m_sutUserSideSingleProducer.getSubscriptionState();
+
+    EXPECT_THAT(subscriptionState, Eq(iox::SubscribeState::UNSUBSCRIBE_REQUESTED));
+}
+
+TEST_F(SubscriberPort_test, ackInUnsubscribeRequestedResultsInNotSubscribed)
+{
+    m_sutUserSideSingleProducer.subscribe();
+    m_sutRouDiSideSingleProducer.getCaProMessage(); // only RouDi changes state
+    iox::capro::CaproMessage caproMessage(iox::capro::CaproMessageType::ACK,
+                                          iox::capro::ServiceDescription("x", "y", "z"));
+    m_sutRouDiSideSingleProducer.dispatchCaProMessage(caproMessage);
+    m_sutUserSideSingleProducer.unsubscribe();
+    m_sutRouDiSideSingleProducer.getCaProMessage(); // only RouDi changes state
+    m_sutRouDiSideSingleProducer.dispatchCaProMessage(caproMessage);
+
+    const auto subscriptionState = m_sutUserSideSingleProducer.getSubscriptionState();
+
+    EXPECT_THAT(subscriptionState, Eq(iox::SubscribeState::NOT_SUBSCRIBED));
+}
+
+TEST_F(SubscriberPort_test, nackInUnsubscribeRequestedResultsInNotSubscribed)
+{
+    m_sutUserSideSingleProducer.subscribe();
+    m_sutRouDiSideSingleProducer.getCaProMessage(); // only RouDi changes state
+    iox::capro::CaproMessage caproMessage(iox::capro::CaproMessageType::ACK,
+                                          iox::capro::ServiceDescription("x", "y", "z"));
+    m_sutRouDiSideSingleProducer.dispatchCaProMessage(caproMessage);
+    m_sutUserSideSingleProducer.unsubscribe();
+    m_sutRouDiSideSingleProducer.getCaProMessage(); // only RouDi changes state
+    caproMessage.m_type = iox::capro::CaproMessageType::NACK;
+    m_sutRouDiSideSingleProducer.dispatchCaProMessage(caproMessage);
+
+    const auto subscriptionState = m_sutUserSideSingleProducer.getSubscriptionState();
+
+    EXPECT_THAT(subscriptionState, Eq(iox::SubscribeState::NOT_SUBSCRIBED));
+}
+
+TEST_F(SubscriberPort_test, invalidMessageResultsInError)
```

Comment:
I think here are some cases missing. 
- What happens when RouDi breaks the protocol and replies on a subscribe with a PING or FIND?

Reply:
will check

Reply:
All transitions of the state machine are tested and the case that a (somehow arbitrary) CaPro message is provided by RouDi which breaks the protocol. What I identified is the case when ACK or NACK are received but we do not wait for them. With these additional two test cases we should have coverage of all statements and branches